Add Watchdog HAL & Driver to Mbed OS

donatieng opened the pull request:
### Description

This PR brings Watchdog functionality from the `feature-watchdog` branch to `master`.

Highlights include:
* Watchdog Driver (Start with timeout, Kick)
* ResetReason Driver (Fetch why the platform was reset based on underlying hardware registers)
* HAL APIs
* HAL Tests
* Platform implementations

### Pull request type

    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [X] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

Commit:
Add Watchdog HAL API specification headers

HAL watchdog functionality will be implemented as two separate APIs. The reset
reason API allows a user to detect the last system reset reason to identify if
a Watchdog was triggered. The Watchdog API allows configuring and updating
Watchdog timers on all boards. This commit defines the headers.

Commit:
Add K64F reset reason reference implementation

Commit:
Add simple watchdog reference implementation for K64F

- Add preprocessor guard to watchdog api that errors if the reset reason api is not also implemented
- Add RESET_REASON and WATCHDOG to K64F targets.json
- Add watchdog reference implementation

Commit:
Fix watchdog API issues

- Fix typo in module comment
- Redefine the default system behaviour in sleep mode
- Guard K64F enableWait flag
- Remove bit shifts from reset reason enum

fkjagodzinski committed:
Tests: HAL API: Watchdog: Add tests

Commit:
Add Reset Reason platform API

Commit:
Add reset reason reference implementation STM32

Commit:
Add function to fetch platform specific reset reason register values

Commit:
Add independent watchdog reference implementation for STM32

Commit:
Amend reset reason driver API
- Change API to match C++ API throughout drivers
- Amend HAL API documentation to be more specific

Commit:
Add Watchdog driver API

fkjagodzinski committed:
Tests: HAL API: Reset_reason: Add tests

Commit:
Remove window and sleep mode options for watchdog API

fkjagodzinski committed:
Tests: HAL API: Watchdog: Add dev reset tests

Commit:
Add missing license headers

Commit:
Move watchdog parameter validation into the driver layer

Commit:
Fix Watchdog::stop on K64F target

WDOG_Disable will suspend the watchdog until a reset is triggered instead of stopping it altogether. Deinit will disable it until it is reinitialized.

fkjagodzinski committed:
K64F: watchdog HAL: Fix max_timeout.

Limit MAX_TIMEOUT_MS to UINT32_MAX.
Improve MAX_TIMEOUT_MS handling.
Fix hal_watchdog_get_platform_features() returning
.max_timeout = 0xfffffff8 instead of 0xffffffff.

fkjagodzinski committed:
Tests: HAL API: Watchdog: Add time accuracy tests

Commit:
Add RESET_REASON and WATCHDOG definitions to hexiwear build target

Commit:
Rename watchdog.c -> watchdog_api.c to prevent name collision with Platform API

fkjagodzinski committed:
Tests: Drivers: Reset_reason: Add tests

fkjagodzinski committed:
Tests: Drivers: Watchdog: Add tests

bulislaw committed:
Add definition of reset reason and watchdog doxy groups

fkjagodzinski committed:
STM32F4: watchdog HAL: Fix issues found with tests

Fix WATCHDOG_STATUS_INVALID_ARGUMENT for timeout values from 1 ms to
407 ms (fix inability to set prescaler bits to zero).
Fix timeout <-> IWDG registers conversions.
Fix read & write access to IWDG_PR and IWDG_RLR registers.
Fix LSI RC frequency setting.
Limit MAX_TIMEOUT_MS to UINT32_MAX.

fkjagodzinski committed:
K64F: watchdog HAL: Fix init() and stop()

Added a missing wait for the WCT window end. Without it, consecutive
init() or stop() calls were ignored.

fkjagodzinski committed:
STM: HAL: Reset_reason: Correct return values

bulislaw committed:
Add RESET_REASON and WATCHDOG to doxygen options

bulislaw committed:
Document reset reason enum

fkjagodzinski committed:
Tests: HAL: Watchdog: Fix timing accuracy test

Relax time measurement margins for UBLOX_EVK_ODIN_W2.

bulislaw committed:
Disable Watchdog on Odin as it fails intermittently

jeromecoutant committed:
STM32 Watchdog : move API file to STM family level

Commit:
Add Watchdog and ResetReason headers to mbed.h

Adding headers to mbed.h will prevent you having to manually include
them when using the drivers in an application.

jeromecoutant committed:
STM32 WATCHDOG : enable STM32 family

stevew817 committed:
Add support for watchdog on Silicon Labs devices

Commit:
Add text to watchdog documentation indicating that it will continue to run in sleep modes

jeromecoutant committed:
STM32 WATCHDOG : compilation issue with typed define

stevew817 committed:
Add implementation of reset reason on Silicon Labs parts

jeromecoutant committed:
STM32 WATCHDOG : update STM32L0 HAL_IWDG_Init to a newest version

jeromecoutant committed:
STM32 WATCHDOG : use ST HAL in order to make code commun for all STM32

Commit:
NCS36510: Added watchdog implementation.

jeromecoutant committed:
STM32 WATCHDOG : increase timeout value

Commit:
fix bugs introduced on rebase

ccli8 committed:
[Nuvoton] Support reset reason

ccli8 committed:
[Nuvoton] Support watchdog timer

ccli8 committed:
[NUC472] Fix WDT driver in BSP

ccli8 committed:
Enlarge wait time for flushing serial buffer in Greentea test code

Original wait time is not enough for UART to flush out before deepsleep/reset on some targets.

ccli8 committed:
Fix NUMAKER_PFM_NANO130 cannot pass mbed_hal/watchdog_reset test

On NUMAKER_PFM_NANO130 target, WDT's clock source is fixed to LIRC, which is much
less accurate than other targets. Enlarge delta define to pass this test.

Commit:
Added reset_reason feature for TMPM066 & TMPM46B

Commit:
Added reset_reason feature for TMPM3H6 & TMPM4G9

Commit:
test(api-watchdog): Fix errors in compiling watchdog greentea tests

The latest rebase of the watchdog feature branch introduced errors in compiling
watchdog tests due to missing headers.

- Watchdog HAL API test
  Include missing header files to main.cpp (mbed_wait_api.h, stdlib.h).

- Watchdog HAL API timing test
  Include missing header files to main.cpp (us_ticker_api.h).

- Watchdog Driver API test
  Include missing header files to main.cpp (mbed_wait_api.h, stdlib.h).
  Inject mbed namespace into main.cpp to fix Watchdog name resolution.

Commit:
fix(hal-watchdog): Add feature guards to K64F watchdog implementation

0xc0170 commented:
>Conflicting files
targets/targets.json

Needs resolving the conflict

0xc0170 commented on TESTS/mbed_drivers/watchdog/main.cpp:
```diff
@@ -0,0 +1,226 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018 ARM Limited
```
can we add SPDX identifier to new files

0xc0170 commented on TESTS/host_tests/reset_reason.py:
```diff
@@ -0,0 +1,162 @@
+"""
+mbed SDK
+Copyright (c) 2017 ARM Limited
```
2018 year fix

line 2 - MBED SDK removed from these files

0xc0170 reviewed (dismissed):
> travis-ci/astyle — Passed, 685 files (+140 files)

Also, needs rebase to get the latest travis configuration and fixing coding style

Commit:
Update license headers

Commit:
Apply new astyle config style

donatieng commented:
@0xc0170 branch is rebased, license headers updated and 'astyled'

0xc0170 commented:
> @0xc0170 branch is rebased, license headers updated and 'astyled'

Tests are still failing, please review the travis

Commit:
Fix more astyle failures

donatieng commented:
Pfew, astyle is happy!

jeromecoutant commented on targets/targets.json:
```diff
@@ -2200,7 +2206,7 @@
         "detect_code": ["0720"],
         "macros_add": ["USB_STM_HAL", "USBHOST_OTHER"],
-        "device_has_add": ["SERIAL_ASYNCH", "SERIAL_FC", "FLASH"],
+        "device_has_add": ["SERIAL_ASYNCH", "SERIAL_FC", "FLASH", "LOWPOWERTIMER"],
```
I think it's a rebase error :-)

jeromecoutant commented on targets/targets.json:
```diff
@@ -1726,13 +1730,15 @@
             "SPI_ASYNCH",
-            "STDIO_MESSAGES"
+            "STDIO_MESSAGES",
+            "WATCHDOG"
```
WATCHDOG and RESET_REASON are added in the STM32 FAMILY level,
so it has to be removed from all STM32 targets after.
Thx

jeromecoutant reviewed (commented):
targets.json needs few updates
Thx

Commit:
Fix rebase issues in targets.json for STM32 targets

donatieng commented:
Thanks for spotting these @jeromecoutant :)

jeromecoutant commented:
ST CI tests:

| target            | platform_name | test suite                        | result | elapsed_time (sec) | copy_method |
|-------------------|---------------|-----------------------------------|--------|--------------------|-------------|
| NUCLEO_F091RC-ARM | NUCLEO_F091RC | tests-mbed_drivers-reset_reason   | OK     | 27.13              | default     |
| NUCLEO_F091RC-ARM | NUCLEO_F091RC | tests-mbed_drivers-watchdog       | OK     | 31.6               | default     |
| NUCLEO_F091RC-ARM | NUCLEO_F091RC | tests-mbed_drivers-watchdog_reset | OK     | 33.59              | default     |
| NUCLEO_F091RC-ARM | NUCLEO_F091RC | tests-mbed_hal-reset_reason       | OK     | 27.21              | default     |
| NUCLEO_F091RC-ARM | NUCLEO_F091RC | tests-mbed_hal-rtc_reset          | OK     | 21.04              | default     |
| NUCLEO_F091RC-ARM | NUCLEO_F091RC | tests-mbed_hal-watchdog           | OK     | 31.79              | default     |
| NUCLEO_F091RC-ARM | NUCLEO_F091RC | tests-mbed_hal-watchdog_reset     | OK     | 33.79              | default     |
| NUCLEO_F091RC-ARM | NUCLEO_F091RC | tests-mbed_hal-watchdog_timing    | OK     | 50.05              | default     |
| NUCLEO_F091RC-ARM | NUCLEO_F091RC | tests-mbed_platform-system_reset  | OK     | 16.67              | default     |
| NUCLEO_F103RB-ARM | NUCLEO_F103RB | tests-mbed_drivers-reset_reason   | OK     | 27.92              | default     |
| NUCLEO_F103RB-ARM | NUCLEO_F103RB | tests-mbed_drivers-watchdog       | OK     | 32.2               | default     |
| NUCLEO_F103RB-ARM | NUCLEO_F103RB | tests-mbed_drivers-watchdog_reset | OK     | 34.65              | default     |
| NUCLEO_F103RB-ARM | NUCLEO_F103RB | tests-mbed_hal-reset_reason       | OK     | 27.97              | default     |
| NUCLEO_F103RB-ARM | NUCLEO_F103RB | tests-mbed_hal-rtc_reset          | OK     | 21.47              | default     |
| NUCLEO_F103RB-ARM | NUCLEO_F103RB | tests-mbed_hal-watchdog           | OK     | 32.55              | default     |
| NUCLEO_F103RB-ARM | NUCLEO_F103RB | tests-mbed_hal-watchdog_reset     | OK     | 34.44              | default     |
| NUCLEO_F103RB-ARM | NUCLEO_F103RB | tests-mbed_hal-watchdog_timing    | FAIL   | 24.38              | default     |
| NUCLEO_F103RB-ARM | NUCLEO_F103RB | tests-mbed_platform-system_reset  | OK     | 17.47              | default     |
| NUCLEO_F207ZG-ARM | NUCLEO_F207ZG | tests-mbed_drivers-reset_reason   | OK     | 26.15              | default     |
| NUCLEO_F207ZG-ARM | NUCLEO_F207ZG | tests-mbed_drivers-watchdog       | OK     | 31.06              | default     |
| NUCLEO_F207ZG-ARM | NUCLEO_F207ZG | tests-mbed_drivers-watchdog_reset | OK     | 32.98              | default     |
| NUCLEO_F207ZG-ARM | NUCLEO_F207ZG | tests-mbed_hal-reset_reason       | OK     | 26.17              | default     |
| NUCLEO_F207ZG-ARM | NUCLEO_F207ZG | tests-mbed_hal-rtc_reset          | OK     | 20.55              | default     |
| NUCLEO_F207ZG-ARM | NUCLEO_F207ZG | tests-mbed_hal-watchdog           | OK     | 31.04              | default     |
| NUCLEO_F207ZG-ARM | NUCLEO_F207ZG | tests-mbed_hal-watchdog_reset     | OK     | 32.73              | default     |
| NUCLEO_F207ZG-ARM | NUCLEO_F207ZG | tests-mbed_hal-watchdog_timing    | OK     | 48.88              | default     |
| NUCLEO_F207ZG-ARM | NUCLEO_F207ZG | tests-mbed_platform-system_reset  | OK     | 15.91              | default     |
| NUCLEO_F303ZE-ARM | NUCLEO_F303ZE | tests-mbed_drivers-reset_reason   | OK     | 27.26              | default     |
| NUCLEO_F303ZE-ARM | NUCLEO_F303ZE | tests-mbed_drivers-watchdog       | OK     | 31.91              | default     |
| NUCLEO_F303ZE-ARM | NUCLEO_F303ZE | tests-mbed_drivers-watchdog_reset | OK     | 33.84              | default     |
| NUCLEO_F303ZE-ARM | NUCLEO_F303ZE | tests-mbed_hal-reset_reason       | OK     | 27.17              | default     |
| NUCLEO_F303ZE-ARM | NUCLEO_F303ZE | tests-mbed_hal-rtc_reset          | OK     | 21.09              | default     |
| NUCLEO_F303ZE-ARM | NUCLEO_F303ZE | tests-mbed_hal-watchdog           | OK     | 31.9               | default     |
| NUCLEO_F303ZE-ARM | NUCLEO_F303ZE | tests-mbed_hal-watchdog_reset     | OK     | 33.92              | default     |
| NUCLEO_F303ZE-ARM | NUCLEO_F303ZE | tests-mbed_hal-watchdog_timing    | OK     | 50.27              | default     |
| NUCLEO_F303ZE-ARM | NUCLEO_F303ZE | tests-mbed_platform-system_reset  | OK     | 16.54              | default     |
| NUCLEO_F446RE-ARM | NUCLEO_F446RE | tests-mbed_drivers-reset_reason   | OK     | 26.6               | default     |
| NUCLEO_F446RE-ARM | NUCLEO_F446RE | tests-mbed_drivers-watchdog       | OK     | 30.83              | default     |
| NUCLEO_F446RE-ARM | NUCLEO_F446RE | tests-mbed_drivers-watchdog_reset | OK     | 32.82              | default     |
| NUCLEO_F446RE-ARM | NUCLEO_F446RE | tests-mbed_hal-reset_reason       | OK     | 26.89              | default     |
| NUCLEO_F446RE-ARM | NUCLEO_F446RE | tests-mbed_hal-rtc_reset          | OK     | 20.72              | default     |
| NUCLEO_F446RE-ARM | NUCLEO_F446RE | tests-mbed_hal-watchdog           | OK     | 31.09              | default     |
| NUCLEO_F446RE-ARM | NUCLEO_F446RE | tests-mbed_hal-watchdog_reset     | OK     | 32.94              | default     |
| NUCLEO_F446RE-ARM | NUCLEO_F446RE | tests-mbed_hal-watchdog_timing    | OK     | 49.35              | default     |
| NUCLEO_F446RE-ARM | NUCLEO_F446RE | tests-mbed_platform-system_reset  | OK     | 15.84              | default     |
| NUCLEO_F767ZI-ARM | NUCLEO_F767ZI | tests-mbed_drivers-reset_reason   | OK     | 26.46              | default     |
| NUCLEO_F767ZI-ARM | NUCLEO_F767ZI | tests-mbed_drivers-watchdog       | OK     | 30.94              | default     |
| NUCLEO_F767ZI-ARM | NUCLEO_F767ZI | tests-mbed_drivers-watchdog_reset | OK     | 32.49              | default     |
| NUCLEO_F767ZI-ARM | NUCLEO_F767ZI | tests-mbed_hal-reset_reason       | OK     | 26.21              | default     |
| NUCLEO_F767ZI-ARM | NUCLEO_F767ZI | tests-mbed_hal-rtc_reset          | OK     | 20.27              | default     |
| NUCLEO_F767ZI-ARM | NUCLEO_F767ZI | tests-mbed_hal-watchdog           | OK     | 30.81              | default     |
| NUCLEO_F767ZI-ARM | NUCLEO_F767ZI | tests-mbed_hal-watchdog_reset     | OK     | 32.45              | default     |
| NUCLEO_F767ZI-ARM | NUCLEO_F767ZI | tests-mbed_hal-watchdog_timing    | OK     | 49.1               | default     |
| NUCLEO_F767ZI-ARM | NUCLEO_F767ZI | tests-mbed_platform-system_reset  | OK     | 15.36              | default     |
| NUCLEO_L073RZ-ARM | NUCLEO_L073RZ | tests-mbed_drivers-reset_reason   | OK     | 29.28              | default     |
| NUCLEO_L073RZ-ARM | NUCLEO_L073RZ | tests-mbed_drivers-watchdog       | OK     | 34.05              | default     |
| NUCLEO_L073RZ-ARM | NUCLEO_L073RZ | tests-mbed_drivers-watchdog_reset | OK     | 36.33              | default     |
| NUCLEO_L073RZ-ARM | NUCLEO_L073RZ | tests-mbed_hal-reset_reason       | OK     | 29.34              | default     |
| NUCLEO_L073RZ-ARM | NUCLEO_L073RZ | tests-mbed_hal-rtc_reset          | OK     | 23.1               | default     |
| NUCLEO_L073RZ-ARM | NUCLEO_L073RZ | tests-mbed_hal-watchdog           | OK     | 34.26              | default     |
| NUCLEO_L073RZ-ARM | NUCLEO_L073RZ | tests-mbed_hal-watchdog_reset     | OK     | 36.2               | default     |
| NUCLEO_L073RZ-ARM | NUCLEO_L073RZ | tests-mbed_hal-watchdog_timing    | OK     | 51.81              | default     |
| NUCLEO_L073RZ-ARM | NUCLEO_L073RZ | tests-mbed_platform-system_reset  | OK     | 18.48              | default     |
| NUCLEO_L152RE-ARM | NUCLEO_L152RE | tests-mbed_drivers-reset_reason   | OK     | 27.35              | default     |
| NUCLEO_L152RE-ARM | NUCLEO_L152RE | tests-mbed_drivers-watchdog       | OK     | 31.72              | default     |
| NUCLEO_L152RE-ARM | NUCLEO_L152RE | tests-mbed_drivers-watchdog_reset | OK     | 33.72              | default     |
| NUCLEO_L152RE-ARM | NUCLEO_L152RE | tests-mbed_hal-reset_reason       | OK     | 27.18              | default     |
| NUCLEO_L152RE-ARM | NUCLEO_L152RE | tests-mbed_hal-rtc_reset          | OK     | 21.06              | default     |
| NUCLEO_L152RE-ARM | NUCLEO_L152RE | tests-mbed_hal-watchdog           | OK     | 32.01              | default     |
| NUCLEO_L152RE-ARM | NUCLEO_L152RE | tests-mbed_hal-watchdog_reset     | OK     | 33.8               | default     |
| NUCLEO_L152RE-ARM | NUCLEO_L152RE | tests-mbed_hal-watchdog_timing    | OK     | 49.78              | default     |
| NUCLEO_L152RE-ARM | NUCLEO_L152RE | tests-mbed_platform-system_reset  | OK     | 16.48              | default     |
| NUCLEO_L476RG-ARM | NUCLEO_L476RG | tests-mbed_drivers-reset_reason   | OK     | 26.65              | default     |
| NUCLEO_L476RG-ARM | NUCLEO_L476RG | tests-mbed_drivers-watchdog       | OK     | 31.3               | default     |
| NUCLEO_L476RG-ARM | NUCLEO_L476RG | tests-mbed_drivers-watchdog_reset | OK     | 33.21              | default     |
| NUCLEO_L476RG-ARM | NUCLEO_L476RG | tests-mbed_hal-reset_reason       | OK     | 26.67              | default     |
| NUCLEO_L476RG-ARM | NUCLEO_L476RG | tests-mbed_hal-rtc_reset          | OK     | 20.62              | default     |
| NUCLEO_L476RG-ARM | NUCLEO_L476RG | tests-mbed_hal-watchdog           | OK     | 31.59              | default     |
| NUCLEO_L476RG-ARM | NUCLEO_L476RG | tests-mbed_hal-watchdog_reset     | OK     | 33.22              | default     |
| NUCLEO_L476RG-ARM | NUCLEO_L476RG | tests-mbed_hal-watchdog_timing    | OK     | 49.59              | default     |
| NUCLEO_L476RG-ARM | NUCLEO_L476RG | tests-mbed_platform-system_reset  | OK     | 16.19              | default     |

jeromecoutant commented:
So only 1 issue:

| NUCLEO_F103RB-ARM | NUCLEO_F103RB | tests-mbed_hal-watchdog_timing    | Timing, 200 ms                         | 0      | 1      | FAIL    | 0.1                |

[1542708967.56][CONN][RXD] >>> Running case #1: 'Timing, 200 ms'...
[1542708967.60][CONN][INF] found KV pair in stream: {{__testcase_start;Timing, 200 ms}}, queued...
mbedgt: :56::FAIL: Values Not Within Delta 55 Expected 200 Was 100

jeromecoutant commented on TESTS/mbed_hal/watchdog_timing/main.cpp:
```diff
@@ -0,0 +1,150 @@
+        return;
+    }
+
+    // Phase 1. -- run the test code.
```
Is it normal that phase 1 is after phase 2 ?

donatieng replied:
I think these get called twice - so phase 1 always happens first (before reset) then phase 2 (post-reset)

fkjagodzinski replied:
Correct. The phase 2 is executed when the host sends a valid time measurement after the device reset.

fkjagodzinski commented on TESTS/host_tests/reset_reason.py:
```diff
@@ -0,0 +1,162 @@
+        super(ResetReasonTest, self).__init__()
+        self.device_has_watchdog = None
+        self.raw_reset_reasons = set()
+        cycle_s = self.get_config_item('program_cycle_s')
```
`forced_reset_timeout` config parameter should be used instead.

fkjagodzinski commented on TESTS/host_tests/sync_on_reset.py:
```diff
@@ -0,0 +1,72 @@
+    def __init__(self):
+        super(SyncOnReset, self).__init__()
+        self.test_case_num = 0
+        cycle_s = self.get_config_item('program_cycle_s')
```
`forced_reset_timeout` config parameter should be used instead.

fkjagodzinski commented on TESTS/host_tests/watchdog_reset.py:
```diff
@@ -0,0 +1,144 @@
+        super(WatchdogReset, self).__init__()
+        self.current_case = TestCaseData(0, CASE_DATA_INVALID)
+        self.__handshake_timer = None
+        cycle_s = self.get_config_item('program_cycle_s')
```
`forced_reset_timeout` config parameter should be used instead.

fkjagodzinski reviewed (commented):
There was an update in #8309 changing the config parameter used in reset test as a delay after the reset and before the `__sync` packet. There is a reference to https://github.com/ARMmbed/htrun/blob/1c570ec77d8c5673dae55dc790302d86d712c36b/mbed_host_tests/__init__.py#L132-L144

Should I prepare a new PR to `feature-watchdog` branch, or update this param after this PR gets merged to master?

bulislaw commented:
What's the status here?

0xc0170 commented:
Is this targeting 5.11 ? I'll label it as such, please let us know.


@jeromecoutant The issue you reported, is it in HAL implementation/test ?

bulislaw commented:
@fkjagodzinski could you implement and push the fixed you proposed?
@0xc0170 ideally 5.11 yes

fkjagodzinski committed:
Tests: Watchdog: Fix config access in the host script

Move the calls to get_config_item() to the setup() method. The config is
not available in the initializer yet.

fkjagodzinski committed:
Tests: Watchdog: Update the reset-to-sync delay

Change the config parameter used as a delay before sending the sync
packet after the device reset in watchdog and reset_reason tests. Use
'forced_reset_timeout' instead of 'program_cycle_s'.

fkjagodzinski commented:
Proposed updates to the host scripts in #8846.

Commit:
Added reset_reason feature for TMPM3HQ

0xc0170 commented:
@fkjagodzinski Can you add your fix here and we start CI asap?

0xc0170 commented:
Starting CI meanwhile to get some initial results

0xc0170 commented:
This is close to ready for integration , waiting for approvals. I'll review soon as well

0xc0170 commented:
@melwee01 / @AnotherButler Can you review these files (drivers/ResetReason.h drivers/Watchdog.h hal/reset_reason_api.h hal/watchdog_api.h ) - new functionality (if it was already done on the feature branch, just approve).

0xc0170 commented:
https://github.com/ARMmbed/mbed-os-5-docs/pull/404 - this is the documentation for this feature. It needs rebase and possibly an update. @scartmell-arm Please look at it

0xc0170 commented:
One PR addition to this feature branch (shall be include or will come later?) #8848

bulislaw commented:
All the PRs need to be engineering ready (marked as "needs: CI") by the end of the day (Austin time). Otherwise it won't make 5.11 and will need to come in the next release (5.12 for features, 5.11.1 for fixes and new platforms).

donatieng commented:
I think this one is good to go!

cmonr commented:
@ARMmbed/mbed-os-core @ARMmbed/mbed-docs If someone could take a look at this today, that would be really awesome.

bulislaw reviewed (approved):
@donatieng please remember about adding driver docs to handbook.

SenRamakri commented on drivers/Watchdog.cpp:
```diff
@@ -0,0 +1,67 @@
+watchdog_status_t Watchdog::start(const uint32_t timeout)
+{
+    if (timeout == 0) {
+        return WATCHDOG_STATUS_INVALID_ARGUMENT;
```
Can we use MBED_ERROR1() here, there is an error code defined for INVALID_ARGUMENT already by mbed-os error coding - https://github.com/ARMmbed/mbed-os/blob/master/platform/mbed_error.h#L721

SenRamakri commented on drivers/Watchdog.cpp:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    if (timeout > max_timeout()) {
+        return WATCHDOG_STATUS_INVALID_ARGUMENT;
```
Please use mbed-os standardized error coding.

SenRamakri commented on hal/watchdog_api.h:
```diff
@@ -0,0 +1,149 @@
+} watchdog_features_t;
+
+
+typedef enum {
```
Maybe we can use Mbed-OS standardized error coding here, which already defines a code for each of these.

SenRamakri commented on targets/TARGET_NUVOTON/TARGET_M451/reset_reason.c:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    if (reset_reason_count == 0) {
+        reset_reason_cast = RESET_REASON_UNKNOWN;
```
If we initialize reset_reason_count with RESET_REASON_UNKNOWN, we can skip the check here.

fkjagodzinski replied:
@SenRamakri I think you made a typo here -- not `reset_reason_count`, but `reset_reason_cast`.

Commit:
Edit ResetReason_tests.h

Make minor grammar edits.

AnotherButler commented on TESTS/mbed_drivers/watchdog/Watchdog_tests.h:
```diff
@@ -0,0 +1,109 @@
+/** Test watchdog can be stopped
+ *
+ * Given a platform with a support for disabling a running watchdog
+ * When watchdog is not running and @a Watchdog::stop() is called
```
Edit for all docs in this PR: Please capitalize "Watchdog" every time it refers to the class.

AnotherButler commented on TESTS/mbed_drivers/watchdog_reset/Watchdog_reset_tests.h:
```diff
@@ -0,0 +1,78 @@
+ */
+void test_deepsleep_reset();
+
+/** Test stopped watchdog can be started again and reset the device
```
Query: What does this mean? That the test stopped, but you can restart watchdog and reset the device? Or that the test stopped watchdog, but you can restart the test and reset the device? Or something else?

melwee01 replied:
Preceding paragraphs are confusing. Should each line be a sentence/sentence fragment? Is it meant to be one thought? Needs punctuation and correct capitalization.

AnotherButler commented on TESTS/mbed_drivers/watchdog_reset/Watchdog_reset_tests.h:
```diff
@@ -0,0 +1,78 @@
+ * Given a device supporting 'disable_watchdog' feature, with watchdog started
+ * When the watchdog is stopped before timeout expires
+ * Then the device is not restarted
+ * When the watchdog is started again and it's timeout expires
```
Please remove the apostrophe from "its"

AnotherButler commented on TESTS/mbed_hal/reset_reason/reset_reason_api_tests.h:
```diff
@@ -0,0 +1,49 @@
+/** Test the Reset Reason HAL API
+ *
+ * Given a device supporting a Reset Reason API
+ * When the device is restarted using various methods
```
Query: What does "various methods" mean? Can we just delete the rest of the line after "restarted"?

AnotherButler commented on TESTS/mbed_hal/watchdog/watchdog_api_tests.h:
```diff
@@ -0,0 +1,102 @@
+ */
+void test_max_timeout_is_valid();
+
+/** Test stopped watchdog can be started again
```
Please see earlier query.

AnotherButler commented on TESTS/mbed_hal/watchdog_reset/watchdog_reset_tests.h:
```diff
@@ -0,0 +1,78 @@
+ */
+void test_deepsleep_reset();
+
+/** Test stopped watchdog can be started again and reset the device
```
Please see earlier query.

AnotherButler commented on TESTS/mbed_hal/watchdog_reset/watchdog_reset_tests.h:
```diff
@@ -0,0 +1,78 @@
+ * Given a device supporting 'disable_watchdog' feature, with watchdog started
+ * When the watchdog is stopped before timeout expires
+ * Then the device is not restarted
+ * When the watchdog is started again and it's timeout expires
```
Please remove apostrophe from "its".

AnotherButler commented on drivers/ResetReason.h:
```diff
@@ -0,0 +1,69 @@
+/** \addtogroup drivers */
+/** ResetReason API. When the system is restarted, the reason for the restart is
+ * contained in the system registers at boot time in a platform specific manner,
+ * this API provides a generic method of fetching the reason for the restart.
```
"...in a platform-specific manner. This API..."

AnotherButler commented on drivers/ResetReason.h:
```diff
@@ -0,0 +1,69 @@
+     *
+     * Platform specific reasons that are not covered by the reset_reason_t enum
+     * will cause the ResetReason::get() function to return
+     * RESET_REASON_PLATFORM. In order to get the actual reason the register
```
Please add comma after "reason"

AnotherButler commented on hal/reset_reason_api.h:
```diff
@@ -0,0 +1,110 @@
+typedef enum {
+    RESET_REASON_POWER_ON,       /**< Set when power is initially applied to the board. The power-on-reset circuit causes a POWER_ON reset when this occurs */
+    RESET_REASON_PIN_RESET,      /**< Set when a reset is triggered by the hardware pin on the board */
+    RESET_REASON_BROWN_OUT,      /**< Triggered when the voltage drops below the low voltage detect (LVD) threshold the system will be held in a reset until the voltage rises above the threshold */
```
"... voltage detect (LVD) threshold. The system..."

AnotherButler commented on hal/reset_reason_api.h:
```diff
@@ -0,0 +1,110 @@
+/** Fetch the reset reason for the last system reset
+ *
+ * This function must return the contents of the system reset reason registers
+ * cast to an appropriate platform independent reset reason. If multiple reset
```
Query: What does this sentence mean? The function must return the contents to a platform-independent reset reason? Or the function returns contents that were cast to a platform-independent reset reason? Or something else?

AnotherButler commented on hal/reset_reason_api.h:
```diff
@@ -0,0 +1,110 @@
+ *
+ * This function must return the contents of the system reset reason registers
+ * cast to an appropriate platform independent reset reason. If multiple reset
+ * reasons are set this function should return RESET_REASON_MULTIPLE. If the
```
Query (throughout the paragraph): It should return that, or it does? Are there times when it should return that but doesn't?

AnotherButler commented on hal/reset_reason_api.h:
```diff
@@ -0,0 +1,110 @@
+ * function should return RESET_REASON_PLATFORM. If no reset reason can be
+ * determined this function should return RESET_REASON_UNKNOWN.
+ *
+ * This function is not idempotent, there is no guarantee that the system
```
Please change the comma to a period or semicolon. Then please fix the confusing double-negative in the next sentence.

AnotherButler commented on hal/reset_reason_api.h:
```diff
@@ -0,0 +1,110 @@
+ * that span multiple registers/addresses the value should be concatenated into
+ * the return type.
+ *
+ * This function is not idempotent, there is no guarantee that the system
```
Please see earlier comment.

AnotherButler commented on hal/watchdog_api.h:
```diff
@@ -0,0 +1,149 @@
+ * which is an embedded peripheral that will reset the system in the case of
+ * system failures or malfunctions.
+ *
+ * The watchdog timer initialises a system timer with a time period specified in
```
Please change "initialises" to "initializes" for U.S. spelling.

AnotherButler commented on hal/watchdog_api.h:
```diff
@@ -0,0 +1,149 @@
+    * setting is platform dependent, to find the maximum value for the current
+    * platform call hal_watchdog_get_features() and check the timeout value
+    * member. The minimum valid value for this setting is 1, attempting to
+    * initialise the watchdog with a timeout of 0ms will return
```
Please change "initialise" to "initialize" for U.S. spelling.

AnotherButler commented on hal/watchdog_api.h:
```diff
@@ -0,0 +1,149 @@
+extern "C" {
+#endif
+
+/** Initialise and start a watchdog timer with the given configuration.
```
Please change "initialise" to "initialize" for U.S. spelling.

AnotherButler reviewed (changes requested):
I've left queries to be addressed. Please note that I think these files require more editing, but in the interest of time, I thought I'd initially address the most confusing issues.

kegilbert commented on doxygen_options.json:
```diff
@@ -4,10 +4,10 @@
-    "INCLUDE_FILE_PATTERNS": "",
-    "PREDEFINED": "DOXYGEN_ONLY DEVICE_ANALOGIN DEVICE_ANALOGOUT DEVICE_CAN DEVICE_CRC DEVICE_ETHERNET DEVICE_EMAC DEVICE_FLASH  DEVICE_I2C DEVICE_I2CSLAVE DEVICE_I2C_ASYNCH DEVICE_INTERRUPTIN DEVICE_ITM DEVICE_LPTICKER DEVICE_PORTIN DEVICE_PORTINOUT DEVICE_PORTOUT DEVICE_PWMOUT DEVICE_RTC DEVICE_TRNG DEVICE_SERIAL DEVICE_SERIAL_ASYNCH DEVICE_SERIAL_FC DEVICE_SLEEP DEVICE_SPI DEVICE_SPI_ASYNCH DE ...
-    "EXPAND_AS_DEFINED": "",
+    "INCLUDE_PATH": "", 
```
Minor nit:
This and the following lines had some white space added (not blocking this review in the interest of time can fix later).

deepikabhavnani commented on drivers/ResetReason.cpp:
```diff
@@ -0,0 +1,47 @@
+    // Call get raw to cache the reset reason before clearing the registers.
+    ResetReason::get_raw();
+
+    hal_reset_reason_clear();
```
As per hal API, reset register `must` be cleared between resets failing to do so may make it difficult to determine
the cause of any subsequent system resets. But the driver does not take care of that,

We should be having destructor+deinit to clear it always. In this case if user does not mandatory call `get` API it will result in incorrect values.

Also thread safety is not taken care of

0xc0170 commented:
@scartmell-arm Will you be able to look at the reviews above? This needs more work

melwee01 commented:
> @melwee01 / @AnotherButler Can you review these files (drivers/ResetReason.h drivers/Watchdog.h hal/reset_reason_api.h hal/watchdog_api.h ) - new functionality (if it was already done on the feature branch, just approve).

I'm taking a look now.

melwee01 commented on TESTS/mbed_drivers/watchdog_reset/Watchdog_reset_tests.h:
```diff
@@ -0,0 +1,78 @@
+ * When the watchdog is kicked before timeout expires
+ * Then the device restart is prevented
+ * When the watchdog is *NOT* kicked again before next timeout expires
+ * Then the device is restarted
```
Same here as above: are these separate sentences or not?

melwee01 commented on TESTS/mbed_drivers/watchdog_reset/main.cpp:
```diff
@@ -0,0 +1,286 @@
+/* mbed Microcontroller Library
```
Mbed

melwee01 commented on TESTS/mbed_hal/reset_reason/main.cpp:
```diff
@@ -0,0 +1,153 @@
+ * Take NUMAKER_PFM_NUC472 as an example:
+ * Its UART peripheral has 16-byte Tx FIFO. With baud rate set to 9600, flush
+ * Tx FIFO would take: 16 * 8 * 1000 / 9600 = 13.3 (ms). So set wait time to
+ * 20ms here for safe.
```
here to be safe.

melwee01 commented on TESTS/mbed_hal/reset_reason/main.cpp:
```diff
@@ -0,0 +1,153 @@
+ *
+ * This should be replaced with a better function that checks if the
+ * hardware buffers are empty. However, such an API does not exist now,
+ * so we'll use the wait_ms() function for now.
```
~we'll~

melwee01 commented on TESTS/mbed_hal/watchdog_reset/main.cpp:
```diff
@@ -0,0 +1,309 @@
+
+/* Flush serial buffer before deep sleep/reset
+ *
+ * Since deepsleep()/reset would shut down the UART peripheral, we wait for some time
```
use or instead of /

bulislaw commented:
The comments need to be addressed ASAP or it will miss 5.11 @donatieng @scartmell-arm

donatieng commented:
Thanks for the reviews - moving to 5.12 after chat with/feedback from @bulislaw @c1728p9 @ChiefBureaucraticOfficer and above in the thread, issues need to be addressed before merging

c1728p9 commented on drivers/Watchdog.cpp:
```diff
@@ -0,0 +1,67 @@
+
+namespace mbed {
+
+watchdog_status_t Watchdog::start(const uint32_t timeout)
```
This should be thread or interrupt safe, but is neither currently.

c1728p9 commented on drivers/Watchdog.cpp:
```diff
@@ -0,0 +1,67 @@
+
+void Watchdog::kick()
+{
+    hal_watchdog_kick();
```
If multiple watchdog object are created they will all be kicking the same watchdog. You won't be able to protect multiple threads with a watchdog, as several threads may be deadlocked but if at least one is still kicking the the watchdog will not cause a reset. This problem was unresolved in the [previous discussion](https://github.com/ARMmbed/mbed-os/pull/5488/files#r153997383). CC @SenRamakri

c1728p9 commented on hal/reset_reason_api.h:
```diff
@@ -0,0 +1,110 @@
+ *
+ * @return enum containing the last reset reason for the board.
+ */
+reset_reason_t hal_reset_reason_get(void);
```
This may be something that needs to be addressed at the driver layer, but what are the guidelines for using this? If the reset flags are cleared on first read, then applications with a bootloader, like cloud client, may never know the reset reason if the bootloader uses the watchdog.

Should there be a recommendation like this `shouldn't be used from bootloaders` or `must always be called on boot from the main application`? Alternatively should this be handled at the driver layer and stored into persistent ram, so bootloader and applications are free to use the driver API? Previous discussion [here](https://github.com/ARMmbed/mbed-os/pull/5488#discussion_r153566734). @SenRamakri @deepikabhavnani what are your thoughts on this?

deepikabhavnani replied:
Agree with @c1728p9 - Reset reason should not be cleared by HAL, also if it is cleared by underlying hardware (clear on read) should be taken care by driver. Drivers are thread safe, so makes more sense in having storage in driver

deepikabhavnani commented on drivers/Watchdog.h:
```diff
@@ -0,0 +1,108 @@
+
+#include "watchdog_api.h"
+
+#include <cstdio>
```
```suggestion
#include <stdio.h>
```

deepikabhavnani commented on drivers/Watchdog.h:
```diff
@@ -0,0 +1,108 @@
+     *                 WATCHDOG_NOT_SUPPORTED if one of the enabled input
+     *                 parameters is not supported by the current platform.
+     */
+    watchdog_status_t start(const uint32_t timeout);
```
Can we have multiple watchdog timers?

deepikabhavnani commented on drivers/Watchdog.h:
```diff
@@ -0,0 +1,108 @@
+     *
+     * If the watchdog timer is not currently running this function does nothing
+     */
+    void kick();
```
Refreshing of watchdog timer is left to the application? What if application enters sleep / deep sleep?
Should core internally call `kick` for initialized watchdog instead of relying on application?

deepikabhavnani commented on drivers/Watchdog.h:
```diff
@@ -0,0 +1,108 @@
+     *
+     * @return Reload value for the watchdog timer in milliseconds.
+     */
+    uint32_t reload_value() const;
```
Can we rename this to `get_reload_value` - It looks like application is trying to `reload the value` and not get the reload value?

deepikabhavnani commented on drivers/Watchdog.h:
```diff
@@ -0,0 +1,108 @@
+     * @return Maximum refresh value supported by the watchdog for the current
+     *         platform in milliseconds
+     */
+    static uint32_t max_timeout();
```
Should we allow to `set` along with this get - `max_timeout` ?

deepikabhavnani reviewed (changes requested):
I have few concerns related to driver side implementation of `watchdog` and `reset_reason`. We will be supporting single instance of watchdog timer or multiple timers are allowed which does not make sense, but will like to understand the goal? 
Also application should be allowed to sleep for longer sleep cycles, and it might not trigger watchdog in sleep. `kick` call should be handled by core internally as part of sleep handling, to make sure the core is up and running and we not stalled, instead relying on application counting delay times per loop to make sure `kick` is called.

Time in this main while loop can vary vastly based on application, and it cannot be deterministic
```

 * while (true) {
 *    watchdog.kick();
 *
 *    // Application code
 * }
```

cmonr commented:
@donatieng When you get a chance, this still needs work.

cmonr commented:
...Huh. @donatieng should this PR be closed, since it seems that work is still needed?

Since https://github.com/ARMmbed/mbed-os/pull/8989 has passed, a force push will be coming in to update the branch, but pretty sure that action will blow away this PR as well.

ghost commented:
> ...Huh. @donatieng should this PR be closed, since it seems that work is still needed?

The issues are currently being addressed, but will likely go in a different PR. I think this one can be closed.
